crates/oxc_linter/src/service.rs

```diff
@@ -69,17 +179,106 @@ impl LintService {
         };
 
         let program = allocator.alloc(ret.program);
-        let semantic_ret = SemanticBuilder::new(source_text, source_type)
+
+        // Build the module record to unblock other threads from waiting for too long.
+        // The semantic model is not built at this stage.
+        let semantic_builder = SemanticBuilder::new(source_text, source_type)
             .with_trivias(ret.trivias)
             .with_check_syntax_error(check_syntax_errors)
-            .with_module_record_builder(true)
-            .build(program);
+            .build_module_record(program);
+        let module_record = semantic_builder.module_record();
+
+        if self.linter.options().import_plugin {
+            self.module_map
+                .insert(path.to_path_buf().into_boxed_path(), Arc::clone(&module_record));
+            self.update_cache_state(path);
+
+            // Stop if the current module is not marked for lint.
+            if !self.paths.contains(path) {
+                return vec![];
+            }
+
+            let cwd = &self.cwd;
+            let canonicalized = self.cwd.join(path);
+            let dir = canonicalized.parent().unwrap();
+
+            // Retrieve all dependency modules from this module.
+            module_record
```

Comment:
Could you tell me why we need retrieve all dependency modules? It will run linter for all of the them?

Reply:
Because the import plugin is mostly matching import and exports.